Documentation/proxy.md

```diff
@@ -6,9 +6,34 @@ etcd currently supports two proxy modes: `readwrite` and `readonly`. The default
 
-The member list used by proxy consists of all client URLs advertised within the cluster, as specified in each members' `-advertise-client-urls` flag. If this flag is set incorrectly, requests sent to the proxy are forwarded to wrong addresses and then fail. Including URLs in the `-advertise-client-urls` flag that point to the proxy itself, e.g. http://localhost:2379, is even more problematic as it will cause loops, because the proxy keeps trying to forward requests to itself until its resources (memory, file descriptors) are eventually depleted. The fix for this problem is to restart etcd member with correct `-advertise-client-urls` flag. After client URLs list in proxy is recalculated, which happens every 30 seconds, requests will be forwarded correctly.
-
-You should start the cluster machines first and then start the proxy as the proxy gets its list from reachable machines. If the proxy starts and is unable to reach a machine it will refresh aggressively. Once the machine list is non-empty the proxy will refresh at the original duration.
+The member list used by an etcd proxy consists of all client URLs advertised in the cluster. These client URLs are specified in each etcd cluster member's `advertise-client-urls` option.
+
+An etcd proxy examines several command-line options to discover its endpoints (peer URLs). In order of precedence, these options are `discovery`, `discovery-srv`, and `initial-cluster`. `initial-cluster` option takes one or more etcd peer URLs as arguments to discover the actual etcd cluster.
+
+After establishing a list of peer URLs in this manner, the proxy retrieves the list of client URLs from the first reachable peer. These client URLs are specified by the `advertise-client-urls` option to etcd, which takes one or more etcd client service endpoint URLs as arguments. The proxy then connects to the first reachable etcd member (we call it peer between an etcd member and another member, and inside raft cluster.) every thirty seconds to refresh this list of endpoints.
+
+While etcd proxies therefore do not need to be given the `advertise-client-urls` option, as they retrieve it from the cluster, this implies that the `initial-cluster` must be set correctly for every proxy (non-member) and `advertise-client-urls` options must be set correctly for every non-proxy, first-order cluster member (please check this [example configuration](https://github.com/coreos/etcd/blob/master/Procfile)). Otherwise, requests to any etcd proxy would be forwarded improperly. Take special care not to give URL arguments to the `advertise-client-urls` option that point to the proxy itself, as such a configuration will cause the proxy to enter a suicidal loop, forwarding requests to itself until resources are exhausted. To correct either case, stop etcd and restart it with the correct URLs.
+
+To summarize the logic:
+
+1. Execute the following steps in order until the *peer-urls* are known:
```

Comment:
1. etcd proxies execute the following steps in order until the cluster *peer-urls* are known: